File: NiftyOptionCalculator-iOS/NiftyOptionCalculator/ViewModels/PnLSimulatorViewModel.swift
import Foundation
import Combine

// MARK: - Simulation Supporting Types

struct SimulationLeg: Identifiable {
    let id = UUID()
    let spotPrice: Double
    let strikePrice: Double
    let optionType: OptionType
    let entryPremium: Double
    let iv: Double              // decimal (e.g. 0.15)
    let daysToExpiry: Int
    let lotSize: Int
    let quantity: Int
}

struct LegSimResult: Identifiable {
    let id = UUID()
    let leg: SimulationLeg
    let newPrice: Double
    let pnl: Double
    let pnlPercent: Double
    let newDelta: Double
    let newGamma: Double
    let newTheta: Double
    let newVega: Double
}

// PayoffPoint defined in Models/Strategy.swift

// MARK: - PnLSimulatorViewModel

@MainActor
final class PnLSimulatorViewModel: ObservableObject {

    // MARK: - Published Properties (Sliders)

    @Published var spotChangePercent: Double = 0     // range -10...10
    @Published var daysElapsed: Double = 0           // range 0...maxDTE
    @Published var ivChangePercent: Double = 0       // range -30...30

    // MARK: - Published Results

    @Published var totalPnl: Double = 0
    @Published var totalPnlPercent: Double = 0
    @Published var legResults: [LegSimResult] = []
    @Published var payoffPoints: [PayoffPoint] = []

    // MARK: - Legs

    var legs: [SimulationLeg] = []

    var maxDTE: Int {
        legs.map(\.daysToExpiry).max() ?? 30
    }

    // MARK: - Private

    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init() {
        setupAutoSimulation()
    }

    // MARK: - Public Methods

    func addLeg(_ leg: SimulationLeg) {
        legs.append(leg)
        simulate()
    }

    func removeLeg(at index: Int) {
        guard index >= 0 && index < legs.count else { return }
        legs.remove(at: index)
        simulate()
    }

    /// Convenience for setting up a single option simulation.
    func setupSingleOption(
        spotPrice: Double,
        strikePrice: Double,
        optionType: OptionType,
        premium: Double,
        iv: Double,
        dte: Int,
        lotSize: Int,
        quantity: Int = 1
    ) {
        legs.removeAll()
        let leg = SimulationLeg(
            spotPrice: spotPrice,
            strikePrice: strikePrice,
            optionType: optionType,
            entryPremium: premium,
            iv: iv,
            daysToExpiry: dte,
            lotSize: lotSize,
            quantity: quantity
        )
        legs.append(leg)
        // Reset sliders
        spotChangePercent = 0
        daysElapsed = 0
        ivChangePercent = 0
        simulate()
    }

    /// Core simulation logic, called whenever sliders change.
    func simulate() {
        guard !legs.isEmpty else {
            totalPnl = 0
            totalPnlPercent = 0
            legResults = []
            payoffPoints = []
            return
        }

        var results: [LegSimResult] = []
        var totalPnlAccum: Double = 0
        var totalInvestment: Double = 0

        for leg in legs {
            let newSpot = leg.spotPrice * (1.0 + spotChangePercent / 100.0)
            let newIV = max(0.01, min(5.0, leg.iv * (1.0 + ivChangePercent / 100.0)))
            let remainingDays = max(leg.daysToExpiry - Int(daysElapsed), 0)
            let T = Double(remainingDays) / 365.0

            let newPrice: Double
            let greeks: GreeksResult

            if T > 0 {
                // Use BlackScholesEngine to compute new option price
                switch leg.optionType {
                case .call:
                    newPrice = BlackScholesEngine.calculateCallPrice(
                        spotPrice: newSpot,
                        strikePrice: leg.strikePrice,
                        timeToExpiry: T,
                        riskFreeRate: 0.065,
                        volatility: newIV
                    )
                    greeks = BlackScholesEngine.calculateCallGreeks(
                        spotPrice: newSpot,
                        strikePrice: leg.strikePrice,
                        timeToExpiry: T,
                        riskFreeRate: 0.065,
                        volatility: newIV
                    )
                case .put:
                    newPrice = BlackScholesEngine.calculatePutPrice(
                        spotPrice: newSpot,
                        strikePrice: leg.strikePrice,
                        timeToExpiry: T,
                        riskFreeRate: 0.065,
                        volatility: newIV
                    )
                    greeks = BlackScholesEngine.calculatePutGreeks(
                        spotPrice: newSpot,
                        strikePrice: leg.strikePrice,
                        timeToExpiry: T,
                        riskFreeRate: 0.065,
                        volatility: newIV
                    )
                }
            } else {
                // Expired - intrinsic value only
                switch leg.optionType {
                case .call:
                    newPrice = max(0, newSpot - leg.strikePrice)
                case .put:
                    newPrice = max(0, leg.strikePrice - newSpot)
                }
                greeks = GreeksResult(delta: 0, gamma: 0, theta: 0, vega: 0, rho: 0)
            }

            let positionSize = Double(leg.quantity) * Double(leg.lotSize)
            let legPnl = (newPrice - leg.entryPremium) * positionSize

            let legPnlPercent: Double
            if leg.entryPremium > 0 {
                legPnlPercent = ((newPrice - leg.entryPremium) / leg.entryPremium) * 100.0
            } else {
                legPnlPercent = 0
            }

            totalPnlAccum += legPnl
            totalInvestment += leg.entryPremium * positionSize

            results.append(LegSimResult(
                leg: leg,
                newPrice: newPrice,
                pnl: legPnl,
                pnlPercent: legPnlPercent,
                newDelta: greeks.delta,
                newGamma: greeks.gamma,
                newTheta: greeks.theta,
                newVega: greeks.vega
            ))
        }

        legResults = results
        totalPnl = totalPnlAccum
        if totalInvestment > 0 {
            totalPnlPercent = (totalPnlAccum / totalInvestment) * 100.0
        } else {
            totalPnlPercent = 0
        }

        // Generate payoff chart points
        generatePayoffChart()
    }

    /// Resets all sliders to their default values.
    func resetSliders() {
        spotChangePercent = 0
        daysElapsed = 0
        ivChangePercent = 0
    }

    // MARK: - Private Methods

    private func setupAutoSimulation() {
        // Debounce slider changes and auto-simulate
        Publishers.CombineLatest3(
            $spotChangePercent,
            $daysElapsed,
            $ivChangePercent
        )
        .debounce(for: .milliseconds(50), scheduler: RunLoop.main)
        .sink { [weak self] _, _, _ in
            self?.simulate()
        }
        .store(in: &cancellables)
    }

    private func generatePayoffChart() {
        guard let firstLeg = legs.first else {
            payoffPoints = []
            return
        }

        let baseSpot = firstLeg.spotPrice
        let rangeMin = baseSpot * 0.90
        let rangeMax = baseSpot * 1.10
        let step = (rangeMax - rangeMin) / 40.0

        var points: [PayoffPoint] = []

        for i in 0...40 {
            let simSpot = rangeMin + step * Double(i)
            var totalPnlAtSpot: Double = 0

            for leg in legs {
                let newIV = max(0.01, min(5.0, leg.iv * (1.0 + ivChangePercent / 100.0)))
                let remainingDays = max(leg.daysToExpiry - Int(daysElapsed), 0)
                let T = Double(remainingDays) / 365.0

                let price: Double
                if T > 0 {
                    switch leg.optionType {
                    case .call:
                        price = BlackScholesEngine.calculateCallPrice(
                            spotPrice: simSpot,
                            strikePrice: leg.strikePrice,
                            timeToExpiry: T,
                            riskFreeRate: 0.065,
                            volatility: newIV
                        )
                    case .put:
                        price = BlackScholesEngine.calculatePutPrice(
                            spotPrice: simSpot,
                            strikePrice: leg.strikePrice,
                            timeToExpiry: T,
                            riskFreeRate: 0.065,
                            volatility: newIV
                        )
                    }
                } else {
                    switch leg.optionType {
                    case .call:
                        price = max(0, simSpot - leg.strikePrice)
                    case .put:
                        price = max(0, leg.strikePrice - simSpot)
                    }
                }

                let positionSize = Double(leg.quantity) * Double(leg.lotSize)
                totalPnlAtSpot += (price - leg.entryPremium) * positionSize
            }

            points.append(PayoffPoint(price: simSpot, payoff: totalPnlAtSpot))
        }

        payoffPoints = points
    }
}

File: OptixAndroid/app/src/main/java/com/optix/app/presentation/screens/simulator/PnLSimulatorViewModel.kt
package com.optix.app.presentation.screens.simulator

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.optix.app.core.calculation.BlackScholesEngine
import com.optix.app.core.constants.AppConstants
import com.optix.app.domain.model.Greeks
import com.optix.app.domain.model.OptionType
import com.optix.app.domain.model.PayoffPoint
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

// ──────────────────────────────────────────────
// Data models
// ──────────────────────────────────────────────

data class SimulationLeg(
    val id: String = UUID.randomUUID().toString(),
    val spotPrice: Double,
    val strikePrice: Double,
    val optionType: OptionType,
    val position: LegDirection,
    val entryPremium: Double,
    val iv: Double,              // decimal e.g. 0.15 for 15%
    val daysToExpiry: Int,
    val lotSize: Int,
    val quantity: Int = 1
)

enum class LegDirection { BUY, SELL }

data class LegSimResult(
    val id: String = UUID.randomUUID().toString(),
    val leg: SimulationLeg,
    val newPrice: Double,
    val pnl: Double,
    val pnlPercent: Double,
    val newDelta: Double,
    val newGamma: Double,
    val newTheta: Double,
    val newVega: Double
)

// ──────────────────────────────────────────────
// UI State
// ──────────────────────────────────────────────

data class PnLSimulatorState(
    // Slider values
    val spotChangePercent: Float = 0f,       // -10..10
    val daysElapsed: Float = 0f,             // 0..maxDTE
    val ivChangePercent: Float = 0f,         // -30..30

    // Legs
    val legs: List<SimulationLeg> = emptyList(),

    // Results
    val totalPnl: Double = 0.0,
    val totalPnlPercent: Double = 0.0,
    val legResults: List<LegSimResult> = emptyList(),
    val payoffPoints: List<PayoffPoint> = emptyList(),
    val breakevens: List<Double> = emptyList(),
    val maxProfit: Double = 0.0,
    val maxLoss: Double = 0.0,

    // Add-leg form
    val showAddLegSheet: Boolean = false,
    val formStrike: String = "",
    val formPremium: String = "",
    val formIV: String = "15",
    val formDTE: String = "7",
    val formLotSize: String = "75",
    val formQuantity: String = "1",
    val formOptionType: OptionType = OptionType.CALL,
    val formDirection: LegDirection = LegDirection.BUY,
    val formSpotPrice: String = "22000"
) {
    val maxDTE: Int get() = legs.maxOfOrNull { it.daysToExpiry } ?: 30
}

// ──────────────────────────────────────────────
// ViewModel
// ──────────────────────────────────────────────

@OptIn(FlowPreview::class)
@HiltViewModel
class PnLSimulatorViewModel @Inject constructor() : ViewModel() {

    private val _state = MutableStateFlow(PnLSimulatorState())
    val state: StateFlow<PnLSimulatorState> = _state.asStateFlow()

    init {
        // Auto-simulate whenever sliders change (debounced)
        viewModelScope.launch {
            snapshotFlow {
                Triple(
                    _state.value.spotChangePercent,
                    _state.value.daysElapsed,
                    _state.value.ivChangePercent
                )
            }
                .debounce(50)
                .collect { simulate() }
        }
    }

    // ── Slider updates ──

    fun setSpotChangePercent(value: Float) {
        _state.update { it.copy(spotChangePercent = value) }
    }

    fun setDaysElapsed(value: Float) {
        _state.update { it.copy(daysElapsed = value) }
    }

    fun setIVChangePercent(value: Float) {
        _state.update { it.copy(ivChangePercent = value) }
    }

    fun resetSliders() {
        _state.update { it.copy(spotChangePercent = 0f, daysElapsed = 0f, ivChangePercent = 0f) }
    }

    // ── Leg management ──

    fun showAddLeg() {
        _state.update { it.copy(showAddLegSheet = true) }
    }

    fun hideAddLeg() {
        _state.update { it.copy(showAddLegSheet = false) }
    }

    fun updateFormStrike(v: String) { _state.update { it.copy(formStrike = v) } }
    fun updateFormPremium(v: String) { _state.update { it.copy(formPremium = v) } }
    fun updateFormIV(v: String) { _state.update { it.copy(formIV = v) } }
    fun updateFormDTE(v: String) { _state.update { it.copy(formDTE = v) } }
    fun updateFormLotSize(v: String) { _state.update { it.copy(formLotSize = v) } }
    fun updateFormQuantity(v: String) { _state.update { it.copy(formQuantity = v) } }
    fun updateFormSpotPrice(v: String) { _state.update { it.copy(formSpotPrice = v) } }
    fun updateFormOptionType(t: OptionType) { _state.update { it.copy(formOptionType = t) } }
    fun updateFormDirection(d: LegDirection) { _state.update { it.copy(formDirection = d) } }

    fun confirmAddLeg() {
        val s = _state.value
        val strike = s.formStrike.toDoubleOrNull() ?: return
        val premium = s.formPremium.toDoubleOrNull() ?: return
        val iv = (s.formIV.toDoubleOrNull() ?: 15.0) / 100.0
        val dte = s.formDTE.toIntOrNull() ?: 7
        val lotSize = s.formLotSize.toIntOrNull() ?: 75
        val quantity = s.formQuantity.toIntOrNull() ?: 1
        val spot = s.formSpotPrice.toDoubleOrNull() ?: 22000.0

        val leg = SimulationLeg(
            spotPrice = spot,
            strikePrice = strike,
            optionType = s.formOptionType,
            position = s.formDirection,
            entryPremium = premium,
            iv = iv,
            daysToExpiry = dte,
            lotSize = lotSize,
            quantity = quantity
        )

        _state.update {
            it.copy(
                legs = it.legs + leg,
                showAddLegSheet = false,
                formStrike = "",
                formPremium = ""
            )
        }
        simulate()
    }

    fun removeLeg(legId: String) {
        _state.update { it.copy(legs = it.legs.filter { leg -> leg.id != legId }) }
        simulate()
    }

    // ── Quick presets ──

    fun setupSingleOption(
        spotPrice: Double,
        strikePrice: Double,
        optionType: OptionType,
        premium: Double,
        iv: Double,
        dte: Int,
        lotSize: Int,
        quantity: Int = 1
    ) {
        val leg = SimulationLeg(
            spotPrice = spotPrice,
            strikePrice = strikePrice,
            optionType = optionType,
            position = LegDirection.BUY,
            entryPremium = premium,
            iv = iv,
            daysToExpiry = dte,
            lotSize = lotSize,
            quantity = quantity
        )
        _state.update {
            it.copy(
                legs = listOf(leg),
                spotChangePercent = 0f,
                daysElapsed = 0f,
                ivChangePercent = 0f,
                formSpotPrice = spotPrice.toString()
            )
        }
        simulate()
    }

    // ── Core simulation ──

    private fun simulate() {
        val s = _state.value
        if (s.legs.isEmpty()) {
            _state.update {
                it.copy(
                    totalPnl = 0.0, totalPnlPercent = 0.0,
                    legResults = emptyList(), payoffPoints = emptyList(),
                    breakevens = emptyList(), maxProfit = 0.0, maxLoss = 0.0
                )
            }
            return
        }

        val results = mutableListOf<LegSimResult>()
        var totalPnlAccum = 0.0
        var totalInvestment = 0.0

        for (leg in s.legs) {
            val dirMultiplier = if (leg.position == LegDirection.BUY) 1.0 else -1.0
            val newSpot = leg.spotPrice * (1.0 + s.spotChangePercent / 100.0)
            val newIV = (leg.iv * (1.0 + s.ivChangePercent / 100.0)).coerceIn(0.01, 5.0)
            val remainingDays = (leg.daysToExpiry - s.daysElapsed.toInt()).coerceAtLeast(0)
            val T = remainingDays / 365.0

            val newPrice: Double
            val greeks: Greeks

            if (T > 0) {
                val isCall = leg.optionType == OptionType.CALL
                val result = BlackScholesEngine.calculateGreeks(
                    spotPrice = newSpot,
                    strikePrice = leg.strikePrice,
                    timeToExpiry = T,
                    riskFreeRate = AppConstants.DEFAULT_RISK_FREE_RATE,
                    volatility = newIV,
                    isCall = isCall
                )
                newPrice = result.optionPrice
                greeks = result.greeks
            } else {
                // Expired — intrinsic value only
                newPrice = when (leg.optionType) {
                    OptionType.CALL -> maxOf(0.0, newSpot - leg.strikePrice)
                    OptionType.PUT -> maxOf(0.0, leg.strikePrice - newSpot)
                }
                greeks = Greeks.EMPTY
            }

            val positionSize = leg.quantity.toDouble() * leg.lotSize.toDouble()
            val legPnl = (newPrice - leg.entryPremium) * positionSize * dirMultiplier
            val legPnlPct = if (leg.entryPremium > 0) {
                ((newPrice - leg.entryPremium) / leg.entryPremium) * 100.0 * dirMultiplier
            } else 0.0

            totalPnlAccum += legPnl
            totalInvestment += leg.entryPremium * positionSize

            results.add(
                LegSimResult(
                    leg = leg,
                    newPrice = newPrice,
                    pnl = legPnl,
                    pnlPercent = legPnlPct,
                    newDelta = greeks.delta * dirMultiplier,
                    newGamma = greeks.gamma,
                    newTheta = greeks.theta,
                    newVega = greeks.vega
                )
            )
        }

        val totalPnlPct = if (totalInvestment > 0) (totalPnlAccum / totalInvestment) * 100.0 else 0.0

        // Generate payoff chart + statistics
        val (points, breakevens, maxProfit, maxLoss) = generatePayoffChart(s)

        _state.update {
            it.copy(
                totalPnl = totalPnlAccum,
                totalPnlPercent = totalPnlPct,
                legResults = results,
                payoffPoints = points,
                breakevens = breakevens,
                maxProfit = maxProfit,
                maxLoss = maxLoss
            )
        }
    }

    private data class PayoffResult(
        val points: List<PayoffPoint>,
        val breakevens: List<Double>,
        val maxProfit: Double,
        val maxLoss: Double
    )

    private fun generatePayoffChart(s: PnLSimulatorState): PayoffResult {
        val firstLeg = s.legs.firstOrNull() ?: return PayoffResult(emptyList(), emptyList(), 0.0, 0.0)
        val baseSpot = firstLeg.spotPrice
        val rangeMin = baseSpot * 0.90
        val rangeMax = baseSpot * 1.10
        val step = (rangeMax - rangeMin) / 50.0

        val points = mutableListOf<PayoffPoint>()
        var maxProfit = Double.NEGATIVE_INFINITY
        var maxLoss = Double.POSITIVE_INFINITY

        for (i in 0..50) {
            val simSpot = rangeMin + step * i
            var totalPnlAtSpot = 0.0

            for (leg in s.legs) {
                val dirMultiplier = if (leg.position == LegDirection.BUY) 1.0 else -1.0
                val newIV = (leg.iv * (1.0 + s.ivChangePercent / 100.0)).coerceIn(0.01, 5.0)
                val remainingDays = (leg.daysToExpiry - s.daysElapsed.toInt()).coerceAtLeast(0)
                val T = remainingDays / 365.0

                val price: Double = if (T > 0) {
                    BlackScholesEngine.calculateOptionPrice(
                        spotPrice = simSpot,
                        strikePrice = leg.strikePrice,
                        timeToExpiry = T,
                        riskFreeRate = AppConstants.DEFAULT_RISK_FREE_RATE,
                        volatility = newIV,
                        isCall = leg.optionType == OptionType.CALL
                    )
                } else {
                    when (leg.optionType) {
                        OptionType.CALL -> maxOf(0.0, simSpot - leg.strikePrice)
                        OptionType.PUT -> maxOf(0.0, leg.strikePrice - simSpot)
                    }
                }

                val positionSize = leg.quantity.toDouble() * leg.lotSize.toDouble()
                totalPnlAtSpot += (price - leg.entryPremium) * positionSize * dirMultiplier
            }

            points.add(PayoffPoint(spotPrice = simSpot, payoff = totalPnlAtSpot))
            if (totalPnlAtSpot > maxProfit) maxProfit = totalPnlAtSpot
            if (totalPnlAtSpot < maxLoss) maxLoss = totalPnlAtSpot
        }

        // Find breakeven crossings (where payoff crosses zero)
        val breakevens = mutableListOf<Double>()
        for (i in 0 until points.size - 1) {
            val p1 = points[i]
            val p2 = points[i + 1]
            if ((p1.payoff <= 0 && p2.payoff >= 0) || (p1.payoff >= 0 && p2.payoff <= 0)) {
                // Linear interpolation
                val ratio = kotlin.math.abs(p1.payoff) / (kotlin.math.abs(p1.payoff) + kotlin.math.abs(p2.payoff))
                val be = p1.spotPrice + ratio * (p2.spotPrice - p1.spotPrice)
                breakevens.add(be)
            }
        }

        return PayoffResult(points, breakevens, maxProfit, maxLoss)
    }

    /** Helper for snapshotFlow of MutableStateFlow */
    private fun <T> snapshotFlow(block: () -> T): Flow<T> {
        return _state.map { block() }.distinctUntilChanged()
    }
}
